transport/x/grpc/util.go

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"go.uber.org/yarpc/internal/procedure"
+)
+
+const defaultMethodName = "__default__"
+
+func procedureNameToServiceNameMethodName(procedureName string) (string, string, error) {
+	// TODO: this is hacky
+	serviceName, methodName := procedure.FromName(strings.Replace(procedureName, "/", "___", -1))
+	if serviceName == "" {
+		return "", "", fmt.Errorf("invalid procedure name: %s", procedureName)
+	}
+	if methodName == "" {
+		methodName = defaultMethodName
+	}
+	// TODO: do we really need to do url.QueryEscape?
+	// Are there consequences if there is a diff from the string and the url.QueryEscape string?
+	return url.QueryEscape(serviceName), url.QueryEscape(methodName), nil
```

Comment:
so I originally wrote the queryEscape to handle the straight-up encoding the "::" in the procedure names, but if we're stripping those out we might just be able to enforce a subset of characters for the URL and return an error if procedure names don't adhere to that.